### PR TITLE
feat: Add Atlantic mock server and dummy verifier for L3 testing

### PR DESCRIPTION
**Pull request by 0xvasanth**
## Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [x] Testing
- [ ] Other (please describe):

## What is the current behavior?

Currently, the orchestrator requires a real Atlantic prover service connection for testing proof generation functionality. This creates several issues:
- Dependency on external services for local development and testing
- Inability to test prover interactions in CI/CD pipelines without real Atlantic credentials
- Difficulty simulating edge cases and failure scenarios in the prover service

Resolves: #NA

## What is the new behavior?

This PR introduces a **Mock Atlantic Server** that simulates the Atlantic prover service for testing and development purposes:

### Key Changes:
- **New Mock Atlantic Server** (`test_utils/crates/mock-atlantic-server/`): A complete mock implementation of the Atlantic API
  - Implements all Atlantic API endpoints (submit job, check status, etc.)
  - Supports configurable behavior including failure rates and processing delays
  - Provides realistic job lifecycle simulation with proper status transitions
  - Includes comprehensive Swagger documentation for API reference
  
- **Orchestrator Integration**: Modified the orchestrator to support mock mode
  - Added `atlantic_mock_fact_hash` configuration option to enable mock mode (set to "true" to enable)
  - When enabled, automatically starts the mock Atlantic server
  - Relaxed parameter validation in mock mode to allow testing without real credentials
  
- **Dummy Verifier Contract**: Added deployment script and contract for testing L3 verification
  - `MockGPSVerifier.sol`: Simple verifier contract for testing
  - `deploy_dummy_verifier.sh`: Script to deploy the mock verifier on local Anvil instance

- **Additional Improvements**:
  - Fixed project name and reorganized Cargo dependencies for better maintainability
  - Enhanced test infrastructure for E2E testing scenarios
  - Added support for simulating various proof verification scenarios

### Benefits:
- Enables comprehensive testing of prover integration without external dependencies
- Allows simulation of various failure scenarios for robust error handling
- Speeds up development by removing network latency to external services
- Enables CI/CD testing of proof generation flows
- Provides a controlled environment for testing L3 proof verification

## Does this introduce a breaking change?

**No**

This change is backward compatible. The mock mode is opt-in via the `atlantic_mock_fact_hash` configuration option. When not enabled, the orchestrator behaves exactly as before, connecting to the real Atlantic service.

### Usage Example:
```bash
# Enable mock mode in orchestrator config
# Set the environment variable or pass as CLI argument
export MADARA_ORCHESTRATOR_ATLANTIC_MOCK_FACT_HASH="true"

# The orchestrator will automatically start the mock server
make run-orchestrator-l3

# Deploy dummy verifier for testing (optional)
./test_utils/scripts/deploy_dummy_verifier.sh
```

### Testing:
- The mock server can be run standalone: `cargo run --bin mock-atlantic-server`
- API documentation available in `test_utils/crates/mock-atlantic-server/swagger.json`
- Supports configurable failure rates and delays for comprehensive testing
- Mock server runs on port 8080 by default (configurable)

### Configuration Options:
- `MADARA_ORCHESTRATOR_ATLANTIC_MOCK_FACT_HASH`: Set to "true" to enable mock mode
- Mock server environment variables:
  - `FAILURE_RATE`: Probability of job failure (0.0 to 1.0)
  - `PROCESSING_DELAY_MS`: Simulated processing time for jobs
  - `PORT`: Server port (default: 8080)

This PR significantly improves the testability of the L3 prover integration and enables local development without external dependencies.

### CODE REVIEW

File: orchestrator/src/core/config.rs [hunk 1]
```diff
@@ -199,13 +200,66 @@ impl Config {
...
+
+                    // Create modified Atlantic params that point to the mock server
+                    let mock_atlantic_params = AtlanticValidatedArgs {
+                        atlantic_api_key: atlantic_params.atlantic_api_key.clone(),
+                        atlantic_service_url: format!("http://127.0.0.1:{}", mock_port)
+                            .parse()
+                            .expect("Failed to parse mock server URL"),
+                        atlantic_rpc_node_url: atlantic_params.atlantic_rpc_node_url.clone(),
+                        atlantic_verifier_contract_address:
+                            "0x007a9a039a9471a7bea2962695cebc03f01a702ffa848763a6749052b0396ccc".to_string(), // Hard-coded for mock mode
```
Comment (Mohiiit) [inline orchestrator/src/core/config.rs]: why is this hardcoded tho? are we sure this would be the address all the time? is this on starknet sepolia? if yes then let's add the link to it as well (voyager link)

File: orchestrator/src/core/config.rs [hunk 2]
```diff
@@ -199,13 +200,66 @@ impl Config {
...
     /// # Returns
     /// * `Box<dyn ProverClient>` - The proving service
-    pub(crate) fn build_prover_service(prover_params: &ProverConfig) -> Box<dyn ProverClient + Send + Sync> {
+    pub(crate) async fn build_prover_service(prover_params: &ProverConfig) -> Box<dyn ProverClient + Send + Sync> {
         match prover_params {
             ProverConfig::Sharp(sharp_params) => Box::new(SharpProverService::new_with_args(sharp_params)),
-            ProverConfig::Atlantic(atlantic_params) => Box::new(AtlanticProverService::new_with_args(atlantic_params)),
+            ProverConfig::Atlantic(atlantic_params) => {
+                // If mock_fact_hash is "true", start the mock Atlantic server
+                if atlantic_params.atlantic_mock_fact_hash.eq("true") {
```
Comment (Mohiiit) [inline orchestrator/src/core/config.rs]: but this is true for L3s only right? when we have L2s mocked, we can still make call to the chain right? or maybe not in case of AR

I think it makes sense to have it for both L2 and L3

File: orchestrator/src/types/params/prover.rs [hunk 3]
```diff
@@ -54,43 +53,80 @@ impl TryFrom<RunCmd> for ProverConfig {
...
+                if is_mock_mode {
+                    // Mock mode: provide sensible defaults for missing parameters
+                    Ok(Self::Atlantic(AtlanticValidatedArgs {
+                        atlantic_api_key: atlantic_args.atlantic_api_key.unwrap_or_else(|| "mock-api-key".to_string()),
+                        atlantic_service_url: atlantic_args
+                            .atlantic_service_url
+                            .unwrap_or_else(|| "http://unused-in-mock-mode.com".parse().unwrap()),
+                        atlantic_rpc_node_url: atlantic_args
+                            .atlantic_rpc_node_url
+                            .unwrap_or_else(|| "http://unused-in-mock-mode.com".parse().unwrap()),
```
Comment (Mohiiit) [inline orchestrator/src/types/params/prover.rs]: url to const?

Comment (0xvasanth) [reply orchestrator/src/types/params/prover.rs]: this is dummy URL

File: orchestrator/src/core/config.rs (hunk 2 above)
Comment (0xvasanth) [reply orchestrator/src/core/config.rs]: Re-frame this in better way with new CLI

File: orchestrator/src/core/config.rs (hunk 1 above)
Comment (0xvasanth) [reply orchestrator/src/core/config.rs]: this code is removed and made better now

File: orchestrator/src/cli/prover/atlantic.rs [hunk 4]
```diff
@@ -23,7 +23,7 @@ pub struct AtlanticCliArgs {
     #[arg(required_if_eq("atlantic", "true"))]
     pub atlantic_rpc_node_url: Option<Url>,
 
-    /// Whether to use mock fact registry.
+    /// Enable mock mode for Atlantic prover (set to "true" to enable).
     #[arg(env = "MADARA_ORCHESTRATOR_ATLANTIC_MOCK_FACT_HASH", long)]
```
Comment (heemankv) [inline orchestrator/src/cli/prover/atlantic.rs:28]: In the earlier implementations, this variable was used to be sent to Atlantic.
Now that we won't be mocking anything to the actual Atlantic service, we can rename this variable while we are re-purposing it 
say : MADARA_ORCHESTRATOR_MOCK_ATLANTIC_SERVICE ?
wdyt @itsparser

File: orchestrator/src/core/config.rs [hunk 5]
```diff
@@ -257,16 +263,52 @@ impl Config {
     ///
     /// # Arguments
     /// * `prover_params` - The proving service parameters
-    /// * `params` - The config parameters
     /// # Returns
     /// * `Box<dyn ProverClient>` - The proving service
-    pub(crate) fn build_prover_service(prover_params: &ProverConfig) -> Box<dyn ProverClient + Send + Sync> {
+    pub(crate) async fn build_prover_service(prover_params: &ProverConfig) -> Box<dyn ProverClient + Send + Sync> {
```
Comment (heemankv) [inline orchestrator/src/core/config.rs]: Unsure as to why we need this `async` ?
What am I missing ?

File: orchestrator/src/core/config.rs [hunk 6]
```diff
@@ -257,16 +263,52 @@ impl Config {
...
+        match prover_params {
+            ProverConfig::Atlantic(atlantic_params) => {
+                // Start mock Atlantic server if flag is enabled and we're in testnet
+                if allow_mock_hash_server && atlantic_params.atlantic_network == "TESTNET" {
+                    info!("Mock Atlantic server flag is enabled, starting mock server...");
+
+                    // Start the mock server in a background task
+                    tokio::spawn(async move {
+                        info!("Starting mock Atlantic server on port 4001");
+                        if let Err(e) = utils_mock_atlantic_server::start_mock_atlantic_server().await {
```
Comment (heemankv) [inline orchestrator/src/core/config.rs:317]: Since we are spawning the service from within code, should we not allow to set the port from CLI as well ?
The user isn't made aware about this port requirement while they are starting the app, 
like they will not know about it because it's not an env
as well as it will not be printed at the start of the orchestrator 
wdyt ?

Comment (heemankv) [reply orchestrator/src/core/config.rs:317]: we can always default to 4001 as you have written

File: orchestrator/src/core/config.rs [hunk 7]
```diff
@@ -257,16 +263,52 @@ impl Config {
...
+                            error!("Failed to start mock Atlantic server: {}", e);
+                        }
+                    });
+
+                    // Give the mock server time to start
+                    tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
```
Comment (heemankv) [inline orchestrator/src/core/config.rs:323]: Nitpick : 
if you have bandwidth, we can use `is-alive` endpoint 
else we can skip since it will only be used for testing, totally your call

File: orchestrator/src/tests/config.rs [hunk 8]
```diff
@@ -395,10 +395,10 @@ pub mod implement_client {
         }
     }
 
-    pub(crate) fn init_prover_client(service: ConfigType, params: &EnvParams) -> Box<dyn ProverClient> {
+    pub(crate) async fn init_prover_client(service: ConfigType, params: &EnvParams) -> Box<dyn ProverClient> {
```
Comment (heemankv) [inline orchestrator/src/tests/config.rs]: if `build_prover_service` will not be async then this will also be removed

File: package.json [hunk 9]
```diff
@@ -1,4 +1,5 @@
 {
+  "name": "madara",
```
Comment (heemankv) [inline package.json:2]: already done in other pr, might face conflicts

File: test_utils/crates/mock-atlantic-server/src/server.rs [hunk 10]
```diff
@@ -314,7 +302,37 @@ pub async fn add_job_handler(
...
+                .iter()
+                .filter(|(_, job)| matches!(job.query.status, AtlanticQueryStatus::Done | AtlanticQueryStatus::Failed))
+                .min_by_key(|(_, job)| job.created_at);
+
+            if let Some((id_to_remove, _)) = oldest_done {
+                let id_to_remove = id_to_remove.clone();
+                jobs.remove(&id_to_remove);
+                debug!("Removed old job {} to maintain memory limit", id_to_remove);
+            } else {
+                // If no completed jobs, remove the oldest job regardless of status
```
Comment (heemankv) [inline test_utils/crates/mock-atlantic-server/src/server.rs:334]: Thinking out loud
Will this not hamper the load with with E2E sends the jobs, 
currently it sends 100 ProofCreation jobs together to Atlantic server !

File: orchestrator/src/cli/prover/atlantic.rs (hunk 4 above)
Comment (0xvasanth) [reply orchestrator/src/cli/prover/atlantic.rs:28]: I believe there’s a slight misunderstanding of the implementation. All the existing code remains unchanged — the only addition here is a new CLI flag that allows users to enable the mock service when needed.

File: test_utils/crates/mock-atlantic-server/src/server.rs (hunk 10 above)
Comment (0xvasanth) [reply test_utils/crates/mock-atlantic-server/src/server.rs:334]: never though of that but this is bottle-necking local memory, so better option is to make it as a env variable of the server which will allow use have the flexibility to customize the resource size

File: package.json (hunk 9 above)
Comment (0xvasanth) [reply package.json:2]: it is gone already when you pull

File: orchestrator/src/core/config.rs (hunk 5 above)
Comment (0xvasanth) [reply orchestrator/src/core/config.rs]: yes that is right was used async in previous code change, will revert this back

File: orchestrator/src/core/config.rs (hunk 7 above)
Comment (0xvasanth) [reply orchestrator/src/core/config.rs:323]: Since this is test infrastructure, the simple 2-second sleep is probably fine - it's predictable and won't hang. The is-alive approach is "nicer" but not crucial for a mock server used only in tests.

File: orchestrator/src/core/config.rs (hunk 6 above)
Comment (0xvasanth) [reply orchestrator/src/core/config.rs:317]: I think adding CLI support here might be unnecessary overhead. If someone needs a custom port, they can already run the server manually, 

`PORT=8080 make run-mock-atlantic-server`

Including this directly in the orchestrator doesn’t add much value and could complicate things unnecessarily.

**Review by Mohiiit — APPROVED** [review]

File: test_utils/crates/mock-atlantic-server/src/server.rs (hunk 10 above)
Comment (heemankv) [reply test_utils/crates/mock-atlantic-server/src/server.rs:334]: Is this converted to env var ?

**Review by heemankv — APPROVED** [review]
LGTM
